[CODE REVIEW] - languagefilter.php

smz opened the pull request:
### Description

This is a code review of the "Language Filter" system plugin.

The main objective was code simplification and optimization.
At the same time some small issues/bugs have been solved.

The major differences can be found in the `parseRule()` method where, I think, the logic is also better documented. `getLanguageCookie()` too has some major mods.

Less important modifications have been introduced in `onUserBeforeSave()`, `onUserAfterSave()`, `onUserLogin()` and `onAfterDispatch()`
### Testing instructions
- Have a multilingual site with at least three languages installed
- Test front-end multilingual functionalities (_including logging in with users that have a specific language assigned_)
- Verify correct browser language detection
- Verify correct menu items and component association
- Verify that everything else is OK, no regression exist and some small glitches are gone
### Potentially solved issues (_please test..._)
- #7047 Cookie-based Redirect to 404s in language filter
- #7419 Wrong language prefix automatically added

smz committed:
Code review

smz commented:
... pinging @infograf768! :smile:

smz committed:
Fixing some Travis nitpicking

... on which I don't even agree so much.

smz commented:
Do we have problems with Travis and PHP 5.5?
~~~
Updating dependencies (including require-dev)
  - Installing squizlabs/php_codesniffer (1.5.6)
    Downloading: connection...    Failed to download squizlabs/php_codesniffer from dist: Could not authenticate against github.com
    Now trying to download from source
  - Installing squizlabs/php_codesniffer (1.5.6)
    Cloning 6f3e42d311b882b25b4d409d23a289f4d3b803d5
~~~

mbabker commented:
No.  Travis has just been really flaky as of late.

smz commented:
OK, Thanks!
Will it retest automatically or have I to make some changes to trigger it?

mbabker commented:
Someone with permission has to reset the job.  I just did it.

smz commented:
Great! Thanks, Michael! :smile:

rdeutz commented:
I did a code restructure on the language plugin a while ago, was not merged (don't ask why), maybe there is something in it you can use #6428

smz commented:
Thanks Robert! I missed that, and I will surely look at it and see if there is anything I can bring in!

As you're familiar with languagefilter.php you'll probably may also give a look at my code and see if it looks all-right to you (*if you have some time, of course...*).

rdeutz commented:
A bit busy atm but sure will do.

smz commented on plugins/system/languagefilter/languagefilter.php:
```diff
@@ -228,49 +230,40 @@ public function postprocessNonSEFBuildRule(&$router, &$uri)
 	 * @param   JRouter  &$router  JRouter object.
 	 * @param   JUri     &$uri     JUri object.
 	 *
-	 * @return  void
```
This was wrong: method returns an array

smz commented on plugins/system/languagefilter/languagefilter.php:
```diff
@@ -278,106 +271,76 @@ public function parseRule(&$router, &$uri)
-		$lang = $uri->getVar('lang', $lang_code);
-
-		if (isset($this->sefs[$lang]))
+		else
```
This `else` condition was missing in original code and we dropped through the non-SEF code also while in SEF mode.

Happily (*or unhappily, depending on the point of view!*) this wasn't apparently causing any issue (*beside some lost CPU cycles*)

Hackwar replied:
No, this "else" is breaking code, since it prevents the language from being overriden by a query parameter. Yes, I think that shouldn't be possible either, but it was a huge discussion when this plugin was refactored by me and I had to remove that. So please spare yourself the hassle and remove it again.

infograf768 commented on plugins/system/languagefilter/languagefilter.php:
```diff
@@ -715,14 +667,33 @@ private function setLanguageCookie($lang_code)
 		}
 
+		// No valid cookie, try browser
+		if (!$lang_code && $this->params->get('detect_browser', 0))
```
this should be 
`// No valid cookie, try browser
		if (!$lang_code && $this->params->get('detect_browser', 0) == 1)

smz replied:
Why, JM? In our code (*the whole Joomla! CMS code*) I'm counting only 71 places where we do this, compared to the gazillion of times we don't (*at least around 700*)...

infograf768 commented:
Although it looks like a simplification, I am not sure that checking browser language in the `private function getLanguageCookie()` is very semantic...

smz commented:
> Although it looks like a simplification, I am not sure that checking browser language in the private function `getLanguageCookie()` is very semantic...

Agreed! :smile:
What could be a better name for that function?
- getPreferredLanguage ()
- guessLanguage()
- getLanguage
- getCookieOrBrowserLanguage()
- getClientLanguage()
- ...................................................

smz commented on plugins/system/languagefilter/languagefilter.php:
```diff
@@ -278,106 +271,76 @@ public function parseRule(&$router, &$uri)
-		// We are called via POST. We don't care about the language
-		// and simply set the default language as our current language.
+		// We are called via POST. We don't care about the language and simply set the default language as our current language.
+		// [Errr... this is not exactly what happens here below!!! TBD verify...]
```
Can someone please shed some light of what we actually should do in case of a POST request?
What I did is exactly what we used to do (*albeit through simplification*), but I'm not sure this is correct.
The comment too doesn't agree with the actual code...

Hackwar replied:
The original code was correct. Your change will break several sites/extensions. Removing the second sentence of the comment would have been correct. There might or might not be a language encoded in the URL or the POST data or in the cookie and they have to be checked one by one.

smz committed:
Something more

- Changed method name from getLanguageCookie() to getClientLanguage()
- Moved the check for existing language folder to the construct
- Introduced new checkLanguage() method to replace some common code

smz commented:
Ouch... I made some mods but Travis got stuck again with PHP 5.5...
Can someone, please, re-submit the job?
Thanks! :smile:

rdeutz commented:
done

smz commented:
Thanks, Robert!

Apparently good 'ol Travis now got stuck on PHP:5.3, but don't worry: I'll have a minor commit for tonight and that will re-trigger it. I'll just be sure to cross my finger while committing... :smile:

smz commented:
**In theory** I have some more small mods to push, but as they are limited to code style and comments fixing, **and** Travis finally got through with all PHP versions, I'll defer those minor mods to a second time, if and when there will be consensus that this is valid code worth merging.

Is that OK?

infograf768 commented:
As it is now, code looks like working for me.

smz committed:
CS, comments and a minor simplification

- Removed a couple of useless parenthesis
- Fixed some comments
- There is no need to check Authorized access levels in
onAfterDispatch() as we have already checked for that in the construct

smz committed:
Fix for rel='alternate' links in non-SEF mode

smz commented:
With latest commit I fix an unreported (*I think*) issue:

While in non-SEF mode there were several cases in which home pages were not detected and as a consequence the `rel='alternate'` links were not generated.

I also modified those links to be of the form `/index.php?lang=sef`, without any redundant reference to item IDs. For consistency, I also plan to apply a similar patch to the "Language Switcher"

I think this PR solves issue #7047 too. If testers will confirm this, that issue could be closed

smz commented:
I just pushed a PR (#7286) that modifies the Language Switcher to better adhere to the `rel=alternate` links generated by this PR...

smz commented:
@infograf768 a question: are there cases in which we want the language switcher to function for languages that do not have an associated home page? If not, I can make further simplifications...

infograf768 commented:
>@infograf768 a question: are there cases in which we want the language switcher filter to function for languages that do not have an associated home page? If not, I can make further simplifications...

No. A content language without a home page is like a bicycle without a saddle...
Will look at your last iteration of the PR as soon as possible, but

>While in non-SEF mode there were several cases in which home pages were not detected and as a consequence the rel='alternate' links were not generated.

I first would have to be able to reproduce this.

smz commented:
Easiest way to reproduce the issue with `rel='alternate'` links in home pages, non-SEF mode, is to go to the naked domain (*default language, cookie language or browser detected language do not matter*): you would not find alternates in there. Now switch to another language and then back to the original one: alternates will be there (*and ugly...*).

smz committed:
Further simplification

- Completely remove languages without home pages from the list of
handled languages
- Some comments

smz commented on plugins/system/languagefilter/languagefilter.php:
```diff
@@ -62,15 +62,19 @@ public function __construct(&$subject, $config)
 			foreach ($this->sefs as $sef => $language)
 			{
+				// Check language access, language is enabled, language folder exists, and language has an Home Page
 				// @todo: In Joomla 2.5.4 and earlier access wasn't set. Non modified Content Languages got 0 as access value
```
Do we still need this "todo" comment?

smz commented:
@infograf768 I forgot to mention that if you test this for fixing the non-SEF home page rel=alternate issue, it would be better if you test it together with #7286 

This is not a requirement, but then there will be to kinds of links to home pages, the "short one" in the rel=alternate links and the "long ones" in the Language Switcher. Those links will be both "seen" by crawlers and I think it will be unpredictable which one will be indexed by search engines.

Having the "short form" indexed would be **a lot** better as this will not change even if you'll change the home pages menu itemid (*or even component*) at a second time.

smz committed:
Ooops! Fixed a BAD mistake

Cut-and-Paste error!!

smz committed:
Looping on lang_codes makes more sense than looping on sefs, here.

infograf768 commented:
> Easiest way to reproduce the issue with rel='alternate' links in home pages, non-SEF mode, is to go to the naked domain (default language, cookie language or browser detected language do not matter): you will not find alternates in there. 

This does not change with your patch.

>Now switch to another language and then back to the original one: alternates will be there (and ugly...).

They are changed indeed but is it B/C ?
Also how is sef concerned by the language switcher when no SEF? 

Looks like we need to add htmlspecialchars here btw:
```
					// Current language link
					case ($i == $this->current_lang):
						$language->link = htmlspecialchars(JUri::getInstance()->toString(array('path', 'query')));
						break;
```
Otherwise we get this for associated items (below for it-IT)
```
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="it-it" lang="it-it" dir="ltr">
<head>
	<meta name="viewport" content="width=device-width, initial-scale=1.0" />
	  <meta http-equiv="content-type" content="text/html; charset=utf-8" />
  <meta name="generator" content="Joomla! - Open Source Content Management" />
  <title>Italiano - Multilingua passo dopo passo</title>
  <link href="/testwindows/trunkgitnew/index.php?option=com_content&amp;view=category&amp;id=32&amp;Itemid=151&amp;lang=it&amp;format=feed&amp;type=rss" rel="alternate" type="application/rss+xml" title="RSS 2.0" />
  <link href="/testwindows/trunkgitnew/index.php?option=com_content&amp;view=category&amp;id=32&amp;Itemid=151&amp;lang=it&amp;format=feed&amp;type=atom" rel="alternate" type="application/atom+xml" title="Atom 1.0" />
  <link href="http://localhost:8888/testwindows/trunkgitnew/index.php?option=com_content&amp;view=category&amp;id=12&amp;lang=en&amp;Itemid=122" rel="alternate" hreflang="en-GB" />
  <link href="http://localhost:8888/testwindows/trunkgitnew/index.php?option=com_content&amp;view=category&amp;id=11&amp;lang=fr&amp;Itemid=121" rel="alternate" hreflang="fr-FR" />
  <link href="http://localhost:8888/testwindows/trunkgitnew/index.php?option=com_content&amp;view=category&amp;id=27&amp;lang=de&amp;Itemid=143" rel="alternate" hreflang="de-DE" />
  <link href="http://localhost:8888/testwindows/trunkgitnew/index.php?option=com_content&view=category&id=32&Itemid=151&lang=it" rel="alternate" hreflang="it-IT" />
  <link href="http://localhost:8888/testwindows/trunkgitnew/index.php?option=com_content&amp;view=category&amp;id=21&amp;lang=es&amp;Itemid=131" rel="alternate" hreflang="es-ES" />
  <link href="http://localhost:8888/testwindows/trunkgitnew/index.php?option=com_content&amp;view=category&amp;id=36&amp;Itemid=180&amp;lang=mk" rel="alternate" hreflang="mk-MK" />
  <link href="http://localhost:8888/testwindows/trunkgitnew/index.php?option=com_content&amp;view=category&amp;id=41&amp;Itemid=192&amp;lang=ta" rel="alternate" hreflang="ta-IN" />
  ```

Now, to be honest: it took ages to get the regressions out of the languagefilter. I am frightened by your changes.

smz commented:
@infograf768 

I can't reproduce the fact that you can't reproduce... :smile: Honestly, I consistently get the issue w/o this patch and consistently don't get it when the patch is applied... That's curious... How can we get out of this?

I don't think we have a B/C commitment for non-SEF URLs (*but I may be wrong...*). In any case I sincerely think that the short-form URL form home pages are a lot preferable for search engines, for the reasons I explained above.

You're right that there is a dis-alignment between current link and "other links" (*the htmlspecialchars thing...*), but I think we should fix it the other way around: it is having HTML entities in URI that is wrong. Unsure about this, but will check...

Changes are always a little bit frightening, but if you give an impartial look at my code you'll see (*or at least that's my opinion*) that it is much more *structured* than the current one and this will be a big plus also for future enhancements/modifications...

mbabker commented:
Depends what the before/after URL is and if it breaks functionality.  If it gave a 200 before it should work after.  See the [B/C statement](http://developer.joomla.org/cms/development-strategy.html#backward_compatibility) for URLs.

smz commented:
@mbabker no functionality is broken here: where you will getting 200 you'll still get 200

@infograf768 I'm chaking the htmlentities thing and indeed it is wrong to have those in URLs. URL encoding is different from HTML enconding and the right function to use in PHP is urlencode() (http://php.net/manual/en/function.urlencode.php)

As the *wrong* URLs comes from JRoute::_ I'm afraid we have a problem there... will check.

infograf768 commented:
> it is having HTML entities in URI that is wrong. Unsure about this, but will check...

As you can see in source for rss, atom, opensearch, form actions, etc., we usually use html entities in urls in Joomla.

smz commented:
yes, but I'm not sure this is correct (per RFCs and w3.org standards)...

smz commented:
from a practical point of view it seems there is no problem: browsers do interpret those `&amp;` and we are using them since ages, but... will check...

smz commented:
hmmm...
~~~
$string = this&that
htmlspecialchars($string) = this&that
htmlentities($string) = this&that
urlencode($string) = this%26that
~~~

smz commented:
It seems there are different pint of view on this:
- commonly ampersand is encoded as `&amp;` in URLs
- according to PHP and other sources it should be encoded as `%26`

For sure a simple `&` is wrong: will fix that (with our current convention of using `&amp;`!

Note that we have this issue in current code too: no big issues but the generated HTML will not pass HTML validators

smz commented:
`&amp;` is OK according to w3.org: http://www.w3.org/TR/html4/appendix/notes.html#h-B.2.2

Actually they recommend that:
> ...HTTP server implementors, and in particular, CGI implementors support the use of ";" in place of "&" to save authors the trouble of escaping "&" characters in this manner

.. but this is of course out of question!

smz committed:
Fix ampersand encoding issue in current language URLs in onAfterDispatch()

smz commented:
Latest commit fix ampersand encoding issue for current language URLs in onAfterDispatch()

I'm anyway dubious if this shouldn't be directly fixed in AbstractUri::toString() and, obviously, assure we don't make a double encoding elsewhere. [*OK: I know the answer: B/C...* :stuck_out_tongue_winking_eye: ]

smz committed:
Final polishing and lacquer (hopefully...)

smz committed:
Currecting a BAD mistake

damn copy-paste...

smz commented:
Added #7419 to the list of the potentially solved issues.

If this is deemed correct, I can also modify the code to always give 404 if (*while in SEF mode and without the option to remove the default language code*) the language code is missing

Hackwar commented on plugins/system/languagefilter/languagefilter.php:
```diff
@@ -707,23 +663,56 @@ private function setLanguageCookie($lang_code)
+	 */
+	private function checkLanguage($lang_code)
+	{
+		return (isset($lang_code) && array_key_exists($lang_code, $this->lang_codes));
```
Sorry, but I fail to see the reason for this function. That is 12 lines of code and comments for ```isset($this->lang_codes[$lang_code])``` It means a function call, etc. Especially the first isset() is useless, since the parameter is not optional.

smz replied:
The parameter is not optional but can be empty! Read the code from where it is called.

smz replied:
any way don't worry this PR is dead.

Hackwar replied:
isset($array[$empty]) will still return false.

Hackwar commented on plugins/system/languagefilter/languagefilter.php:
```diff
@@ -602,25 +554,32 @@ public function onAfterDispatch()
+					&& ($current_link == $active_link
+						|| $current_link == '/index.php?lang=' . $languages[$this->current_lang]->sef
+						|| $current_link . 'index.php' == $active_link
+						|| $current_link == $active_link . '/'));
```
In which cases do you have the situation of line 560 or 562? Both cases should never happen and we should rather try to fix those, than to complicate our code here.

infograf768 commented:
Please do not close, @smz

We found out that the cookie issue also exists when url language code is not removed.

smz committed:
Added Copyright for my code

infograf768 commented:
I have updated https://github.com/joomla/joomla-cms/pull/7427

I also discovered (not patched) at the same time that the alternate were not displayed <b>for the home pages</b> when <b>sef was off </b> when we use short urls of the type `http://mysite.com/index.php?lang=xx` while long home page urls (The ones from the Home menu item) the get the alternate fine.
